### PR TITLE
Support Ruby build for OpenBSD/FreeBSD

### PR DESCRIPTION
**Pull request by frenkel**
These two commits add OpenBSD support, at least for the RubyGem. The first commits does all the local changes and the second commit updates cares to a version that builds on OpenBSD.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                     FILE SIZE
 ++++++++++++++ GROWING                                       ++++++++++++++

 -------------- SHRINKING                                     --------------
  -0.0%    -170 [None]                                        -10.3Ki  -0.2%
      -0.0%    -170 [Unmapped]                                    -10.3Ki  -0.2%
      [DEL]     -17 xdigits.6752                                      -17  [DEL]
      [DEL]     -17 xdigits_l.6808                                    -17  [DEL]
      [DEL]     -17 xdigits_u.6809                                    -17  [DEL]
      [DEL]     -11 digits.6753                                       -11  [DEL]
      [DEL]     -11 digits.6781                                       -11  [DEL]
      [DEL]     -11 digits.6794                                       -11  [DEL]
  -8.1%    -624 third_party/cares/cares/ares_process.c           -624  -8.1%
      [DEL] -1.37Ki process_answer.part.11                        -1.37Ki  [DEL]
     -11.7%    -230 ares__send_query                                 -230 -11.7%
      -6.3%    -102 processfds                                       -102  -6.3%
      [DEL]    -100 socket_write.isra.5                              -100  [DEL]
      [DEL]     -61 socket_recv.isra.1                                -61  [DEL]
      -7.3%     -48 read_udp_packets                                  -48  -7.3%
     -27.6%     -40 [Unmapped]                                        -40 -27.6%
      [DEL]     -39 ares__socket_close                                -39  [DEL]
      -0.9%      -4 configure_socket                                   -4  -0.9%
  -0.7%     -64 third_party/cares/cares/ares_init.c               -64  -0.7%
      -6.4%     -32 ares_dup                                          -32  -6.4%
      -0.3%     -16 ares_init_options.part.4                          -16  -0.3%
      [DEL]     -15 ares_set_socket_functions                         -15  [DEL]
      -0.8%      -1 [Unmapped]                                         -1  -0.8%
  -0.7%      -4 third_party/cares/cares/ares_create_query.c        -4  -0.7%
      -0.7%      -4 ares_create_query                                  -4  -0.7%
  -0.8%      -2 third_party/cares/cares/ares__close_sockets.c      -2  -0.8%
      -0.8%      -2 ares__close_sockets                                -2  -0.8%

  -0.1%    -864 TOTAL                                         -10.9Ki  -0.2%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by apolcyn** [comment]
Ah ok, I'm actually not certain from looking at it, but is this updating c-ares to 1.13.0? c-ares was upgraded to 1.13.0, which I believe is the version adding openbsd support, in #12305. This is on master but hasn't been released yet - it will exist in the grpc ruby gem in the 1.17.0 release.

**Comment by frenkel** [comment]
@apolcyn I took master of grpc and it didn't have a new enough c-ares as submodule (the OpenBSD changes were missing). This commit upgrades it to the first version that does build.

**Comment by frenkel** [comment]
BTW, this fixes #10411

**Comment by mehrdada** [comment]
@frenkel  I just verified and we are on 1.13 release commit of `c-ares` on gRPC `master`. We tend to only depend on release tags on the submodule repos and not any commit point. I see that the commit you pointed at is significantly older than what we are on. Can you ensure that your branch is up to date with gRPC `master`?

**Comment by mehrdada** [comment]
@frenkel  I just checked and your branch is not out of date. Please note that you are not upgrading `c-ares`—the one in use already is newer; did you intend to downgrade it? If the current c-ares does not build on OpenBSD, they should fix it in their repo and release a point release *on top of* their latest release before we can depend on it.

File: src/ruby/ext/grpc/extconf.rb [hunk 1]
```diff
@@ -70,7 +71,8 @@
   puts 'Building internal gRPC into ' + grpc_lib_dir
   nproc = 4
   nproc = Etc.nprocessors * 2 if Etc.respond_to? :nprocessors
-  system("make -j#{nproc} -C #{grpc_root} #{grpc_lib_dir}/libgrpc.a CONFIG=#{grpc_config} Q=")
```
Comment (mehrdada) [inline src/ruby/ext/grpc/extconf.rb]: Probably needs to happen for FreeBSD too?

Comment (frenkel) [reply src/ruby/ext/grpc/extconf.rb]: I'm not sure, I've never used FreeBSD.

**Comment by frenkel** [comment]
You're absolutely right, mix up on my side I guess. I've removed the commit and verified everything still builds correctly. Thanks!

File: src/ruby/ext/grpc/extconf.rb (hunk 1 above)
Comment (mehrdada) [reply src/ruby/ext/grpc/extconf.rb]: My point is, I think the platform-specific code in the Ruby build system should be unified for all BSDs, not just special case OpenBSD.

**Comment by frenkel** [comment]
I understand, but I don't know whether others need this. I'm not sure which
make is default on FreeBSD or NetBSD.

Op 11 okt. 2017 9:33 p.m. schreef "Mehrdad Afshari" <
notifications@github.com>:

*@mehrdada* commented on this pull request.
------------------------------

In src/ruby/ext/grpc/extconf.rb
<https://github.com/grpc/grpc/pull/12923#discussion_r144113005>:

> @@ -70,7 +71,8 @@
   puts 'Building internal gRPC into ' + grpc_lib_dir
   nproc = 4
   nproc = Etc.nprocessors * 2 if Etc.respond_to? :nprocessors
-  system("make -j#{nproc} -C #{grpc_root} #{grpc_lib_dir}/libgrpc.a
CONFIG=#{grpc_config} Q=")

My point is, I think the platform-specific code in the Ruby build system
should be unified for all BSDs, not just special case OpenBSD.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/grpc/grpc/pull/12923#discussion_r144113005>, or mute
the thread

File: src/ruby/ext/grpc/extconf.rb (hunk 1 above)
Comment (mehrdada) [reply src/ruby/ext/grpc/extconf.rb]: If I recall correctly, you had to use `gmake` on FreeBSD as well. Let's run `gmake` for all BSDs.

**Comment by frenkel** [comment]
Sure. I'll update this PR tomorrow.

Op 11 okt. 2017 9:42 p.m. schreef "Mehrdad Afshari" <
notifications@github.com>:

**Comment by krytarowski** [comment]
> I understand, but I don't know whether others need this. I'm not sure which make is default on FreeBSD or NetBSD.

NetBSD uses bmake (BSD make(1)), FreeBSD reuses it from NetBSD as is.

File: src/ruby/ext/grpc/extconf.rb (hunk 1 above)
Comment (krytarowski) [reply src/ruby/ext/grpc/extconf.rb]: NetBSD/FreeBSD require gmake for GNU features.

Comment (frenkel) [reply src/ruby/ext/grpc/extconf.rb]: I've made this so that all the BSD's will use `gmake` in a new commit.

**Comment by frenkel** [comment]
Updated this PR, didn't need to disable `-Werror` anymore and made the use of `gmake` work for all BSD's.

**Comment by mehrdada** [comment]
cc @nicolasnoble for taking a look at changes in `.h` and `.cc` files.

File: src/core/lib/support/time_posix.cc [hunk 2]
```diff
@@ -42,7 +42,7 @@ static struct timespec timespec_from_gpr(gpr_timespec gts) {
   return rv;
 }
 
-#if _POSIX_TIMERS > 0
+#if _POSIX_TIMERS > 0 || defined(__OpenBSD__)
```
Comment (mehrdada) [inline src/core/lib/support/time_posix.cc]: Doesn't OpenBSD set `_POSIX_TIMERS` in its `unistd.h`?

Comment (krytarowski) [reply src/core/lib/support/time_posix.cc]: It sets it as `-1`.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by mehrdada** [comment]
@apolcyn The changes LGTM contingent on tests passing. Can you please take a quick look for an additional sanity check?

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.old: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.opt.new: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.new: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.counters.old: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.old: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.new: 1


[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by apolcyn** [comment]
also lgtm pending tests. thanks for reviews @nicolasnoble and @mehrdada.

I think we should also squash down these commits before merging (though we might want to wait for the current round of tests to complete first).

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_InProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.old: 1


[microbenchmarks] No significant performance differences
```

**Comment by mehrdada** [comment]
Test failures:
macOS, Windows: https://github.com/grpc/grpc/issues/12960
Multilang: https://github.com/grpc/grpc/issues/12886
Interop: https://github.com/grpc/grpc/issues/10763
Sanitizers: https://github.com/grpc/grpc/issues/12892

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by mehrdada** [comment]
Fixed https://github.com/grpc/grpc/issues/10411

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_2_0.opt.new: 1
    bm_fullstack_streaming_ping_pong.BM_StreamingPingPongWithCoalescingApi_MinInProcessCHTTP2_NoOpMutator_NoOpMutator__2M_1_0.counters.old: 1


[microbenchmarks] No significant performance differences
```

**Comment by frenkel** [comment]
Thanks for accepting these changes, although you don't officially support OpenBSD.

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @frenkel isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Pull request merged**